File: telemetry/telemetry/core/timeline/tracing/async_slice.py
# Copyright 2013 The Chromium Authors. All rights reserved.
# Use of this source code is governed by a BSD-style license that can be
# found in the LICENSE file.

import telemetry.core.timeline.tracing.slice as tracing_slice

class AsyncSlice(tracing_slice.Slice):
  ''' A AsyncSlice represents an interval of time during which an
  asynchronous operation is in progress. An AsyncSlice consumes no CPU time
  itself and so is only associated with Threads at its start and end point.
  '''
  def __init__(self, category, name, timestamp, args=None, parent=None):
    super(AsyncSlice, self).__init__(
        category, name, timestamp, args=args, parent=parent)
    self._sub_slices = []
    self.start_thread = None
    self.end_thread = None
    self.id = None

  @property
  def sub_slices(self):
    return self._sub_slices

  def AddSubSlice(self, sub_slice):
    self._sub_slices.append(sub_slice)
    sub_slice.parent = self
    self.children.append(sub_slice)

File: src/tracing/trace_model/async_slice.js
// Copyright (c) 2013 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

'use strict';

base.require('tracing.trace_model.slice');

/**
 * @fileoverview Provides the AsyncSlice class.
 */
base.exportTo('tracing.trace_model', function() {
  /**
   * A AsyncSlice represents an interval of time during which an
   * asynchronous operation is in progress. An AsyncSlice consumes no CPU time
   * itself and so is only associated with Threads at its start and end point.
   *
   * @constructor
   */
  function AsyncSlice(category, title, colorId, start, args) {
    tracing.trace_model.Slice.apply(this, arguments);
  };

  AsyncSlice.prototype = {
    __proto__: tracing.trace_model.Slice.prototype,

    toJSON: function() {
      var obj = new Object();
      var keys = Object.keys(this);
      for (var i = 0; i < keys.length; i++) {
        var key = keys[i];
        if (typeof this[key] == 'function')
          continue;
        if (key == 'startThread' || key == 'endThread') {
          obj[key] = this[key].guid;
          continue;
        }
        obj[key] = this[key];
      }
      return obj;
    },

    id: undefined,

    startThread: undefined,

    endThread: undefined,

    subSlices: undefined
  };

  return {
    AsyncSlice: AsyncSlice
  };
});
